### PR TITLE
Add io.file.read operator

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**New feature**

## Description:

- Adds a new operator `io.file.read` that reads raw contents of a file from a file path given as an input.

### Affected modules and functionalities:

- New operator

### Key points relevant for the review:

- Everything is relevant

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: TODO

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/test/python/io/test_io_file.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import numpy as np
+import nvidia.dali.fn as fn
+import os
+import random
+from nvidia.dali import pipeline_def
+import test_utils
+
+
+data_path = os.path.join(os.environ["DALI_EXTRA_PATH"], "db/single/jpeg/")
```
Comment (klecki) [inline dali/test/python/io/test_io_file.py:23]:
```suggestion
data_path = os.path.join(test_utils.get_dali_extra_path(), "db/single/jpeg/")
```

File: dali/test/python/io/test_io_file.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        "db/single/jpeg/425/abandoned-2179173_1280.jpg",
+        "db/single/jpeg/425/barn-227557_1280.jpg",
+        "db/single/jpeg/425/winter-barn-556696_1280.jpg",
+    ]
+    path_strs = [
+        os.path.join(test_utils.get_dali_extra_path(), random.choice(rel_fpaths))
+        for _ in range(batch_size)
+    ]
+    paths_output = [
+        np.frombuffer(path_str.encode("utf-8"), dtype=np.int8) for path_str in path_strs
```
Comment (klecki) [inline dali/test/python/io/test_io_file.py:50]: Can you maybe test the `u8` encoding as well?

File: dali/operators/io/file/file_read.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      std::memcpy(&filenames_[i][0], filepaths.raw_tensor(i), filename_len);
+      filenames_[i][filename_len] = '\0';
+
+      FileStream::Options opts;
+      opts.use_mmap = !dont_use_mmap_;
+      opts.use_odirect = use_o_direct_;
+      opts.read_ahead = false;
+      files_[i] = FileStream::Open(filenames_[i], opts);
+      output_descs[0].shape.tensor_shape_span(i)[0] = files_[i]->Size();
+    }
```
Comment (klecki) [inline dali/operators/io/file/file_read.cc:53]: Is opening a file a possible bottleneck? I think you may have a bit of expertise in that topic based on the imgcodec work?

Would there be a benefit in running that in a thread pool or would it cause more harm?

File: dali/operators/io/file/file_read.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    return true;
+  }
+
+  void RunImpl(Workspace &ws) override {
+    auto &filenames = ws.Input<CPUBackend>(0);
```
Comment (klecki) [inline dali/operators/io/file/file_read.cc:58]: Unused

File: dali/operators/io/file/file_read.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    auto &file_contents = ws.Output<CPUBackend>(0);
+    auto &tp = ws.GetThreadPool();
+    int nthreads = tp.NumThreads();
```
Comment (klecki) [inline dali/operators/io/file/file_read.cc:61]: Unused

File: dali/operators/io/file/file_read.cc [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  bool dont_use_mmap_;
+  bool use_o_direct_;
+  std::vector<std::string> filenames_;
+  std::vector<std::unique_ptr<FileStream>> files_;
+};
+
+}  // namespace io
+
+DALI_REGISTER_OPERATOR(io__file__Read, io::FileRead, CPU);
+DALI_SCHEMA(io__file__Read)
```
Comment (klecki) [inline dali/operators/io/file/file_read.cc:90]: I just want to keep this in the history of conversation, but I would consider `io.read.file` so the potential "readers" for other containers would all land in one module.

File: dali/operators/io/file/file_read.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    .DocStr(R"(Reads raw file contents.)")
```
Comment (klecki) [inline dali/operators/io/file/file_read.cc:91]: IMO this operator is lacking more docs. I would suggest adding a module docstring for the `io` module (or `io.read` if we were to use my suggestion from above), to highlight the difference between `readers` and `io` operators. You can even specify it for both of the modules, helping the user decide which one to choose: predefined dataset, suitable for regular pipeline or the other used for dynamic file reading.


Also, I would suggest some explanation on how to actually pass the filepath into the operator - list of Python strings which is an obvious choice probably won't work :)

Btw, can we actually pass a constant of Python `str` type to this operator and have it still work? Would our constant promotion handle that correctly? If not, Imo that would be interesting test - just to see how bad the error message is and potentially adjust the code that generates it.

File: dali/operators/io/file/file_read.cc (hunk 3 above)
Comment (jantonguirao) [reply dali/operators/io/file/file_read.cc:53]: There's no such thing as opening the file in S3. However, querying the file size involves a server request, so we might want to parallelize it. Good point.

File: dali/operators/io/file/file_read.cc (hunk 7 above)
Comment (jantonguirao) [reply dali/operators/io/file/file_read.cc:91]: This is the error message:
```
TypeError: when calling operator `nvidia.dali.fn.io.file.read`: expected inputs of type 'DataNode', list of 'DataNode' or convertible to constant nodes. Received input `0` of type 'list'.

```

Comment (jantonguirao) [reply dali/operators/io/file/file_read.cc:91]: automatic promotion from constant string to 1D tensor for bytes, would be interesting, but I see this as definitely out of scope

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**